From 167ca8e74b905ae2eef382837c64692817b72234 Mon Sep 17 00:00:00 2001
From: Matthew Denton <mpdenton@chromium.org>
Date: Mon, 14 Jul 2025 14:56:49 -0700
Subject: [PATCH] [M138] Linux sandbox: Allowlist flags to send* socket syscalls

Many of the flags to send* socket syscalls are not necessary. In
particular, we don't want to allowlist MSG_OOB.

(cherry picked from commit dd434a4960b0372178b3833a644bb2475d7a3e12)

Bug: 428177287
Change-Id: If4086f288586c890714d9283521f9c4f51e2c3f1
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6711812
Commit-Queue: Jorge Lucangeli Obes <jorgelo@chromium.org>
Auto-Submit: Matthew Denton <mpdenton@chromium.org>
Reviewed-by: Jorge Lucangeli Obes <jorgelo@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1483979}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6735344
Commit-Queue: Matthew Denton <mpdenton@chromium.org>
Cr-Commit-Position: refs/branch-heads/7204@{#2052}
Cr-Branched-From: d5de512dc9dc8ddfe4e6d71b0637578bb6158683-refs/heads/main@{#1465706}
---

--- a/sandbox/linux/seccomp-bpf-helpers/baseline_policy.cc
+++ b/sandbox/linux/seccomp-bpf-helpers/baseline_policy.cc
@@ -145,6 +145,10 @@
     return Allow();
   }
 
+  if (SyscallSets::IsSockSendOneMsg(sysno)) {
+    return RestrictSockSendFlags(sysno);
+  }
+
 #if defined(__NR_rseq) && !BUILDFLAG(IS_ANDROID)
   // See https://crbug.com/1104160. Rseq can only be disabled right before an
   // execve, because glibc registers it with the kernel and so far it's unclear
--- a/sandbox/linux/seccomp-bpf-helpers/baseline_policy_unittest.cc
+++ b/sandbox/linux/seccomp-bpf-helpers/baseline_policy_unittest.cc
@@ -571,7 +571,68 @@
   int id;
   setsockopt(fds[0], SOL_SOCKET, SO_DEBUG, &id, sizeof(id));
 }
-#endif
+
+BPF_DEATH_TEST_C(BaselinePolicy,
+                 SendFlagsFiltered,
+                 DEATH_SUCCESS(),
+                 BaselinePolicy) {
+  int fds[2];
+  PCHECK(socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == 0);
+
+  char buf[1] = {'x'};
+// Check allowlisted MSG_DONTWAIT with send. Newer platforms don't have send()
+// anymore.
+#if defined(__NR_send)
+  PCHECK(syscall(__NR_send, fds[0], &buf, 1, MSG_DONTWAIT) != -1);
+#endif  //  defined(__NR_send)
+
+  // Check allowlisted MSG_DONTWAIT with sendto
+  PCHECK(syscall(__NR_sendto, fds[0], &buf, 1, MSG_DONTWAIT, nullptr,
+                 nullptr) != -1);
+
+  // Check allowlisted MSG_DONTWAIT with sendmsg
+  struct msghdr msg = {};
+  PCHECK(syscall(__NR_sendmsg, fds[0], &msg, MSG_DONTWAIT) != -1);
+}
+
+#if defined(__NR_send)
+BPF_DEATH_TEST_C(BaselinePolicy,
+                 SendFlagsFilteredMSG_OOB,
+                 DEATH_SEGV_MESSAGE(GetErrorMessageContentForTests()),
+                 BaselinePolicy) {
+  int fds[2];
+  PCHECK(socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == 0);
+
+  char buf[1];
+  // Specifically disallow MSG_OOB
+  syscall(__NR_send, fds[0], &buf, 1, MSG_OOB);
+}
+#endif  //  defined(__NR_send)
+
+BPF_DEATH_TEST_C(BaselinePolicy,
+                 SendfromFlagsFilteredMSG_OOB,
+                 DEATH_SEGV_MESSAGE(GetErrorMessageContentForTests()),
+                 BaselinePolicy) {
+  int fds[2];
+  PCHECK(socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == 0);
+
+  char buf[1];
+  // Specifically disallow MSG_OOB
+  syscall(__NR_sendto, fds[0], &buf, 1, MSG_OOB, nullptr, nullptr);
+}
+
+BPF_DEATH_TEST_C(BaselinePolicy,
+                 SendmsgFlagsFilteredMSG_OOB,
+                 DEATH_SEGV_MESSAGE(GetErrorMessageContentForTests()),
+                 BaselinePolicy) {
+  int fds[2];
+  PCHECK(socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == 0);
+  // Specifically disallow MSG_OOB
+  struct msghdr msg = {};
+  syscall(__NR_sendmsg, fds[0], &msg, MSG_DONTWAIT | MSG_OOB);
+}
+
+#endif  // !defined(i386)
 
 }  // namespace
 
--- a/sandbox/linux/seccomp-bpf-helpers/syscall_parameters_restrictions.cc
+++ b/sandbox/linux/seccomp-bpf-helpers/syscall_parameters_restrictions.cc
@@ -15,6 +15,7 @@
 #include <sys/prctl.h>
 #include <sys/ptrace.h>
 #include <sys/resource.h>
+#include <sys/socket.h>
 #include <sys/stat.h>
 #include <sys/time.h>
 #include <sys/types.h>
@@ -510,4 +511,36 @@
       .Else(CrashSIGSYS());
 }
 
+SANDBOX_EXPORT bpf_dsl::ResultExpr RestrictSockSendFlags(int sysno) {
+  size_t argIndex;
+  switch (sysno) {
+#if defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+    case __NR_send:
+      argIndex = 3;
+      break;
+#endif
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || \
+    defined(__mips__) || defined(__aarch64__)
+    case __NR_sendto:  // Could specify destination.
+      argIndex = 3;
+      break;
+    case __NR_sendmsg:  // Could specify destination.
+      argIndex = 2;
+      break;
+#endif
+    case __NR_sendmmsg:  // Could specify destination.
+      argIndex = 3;
+      break;
+    default:
+      NOTREACHED();
+  }
+
+  // In particular, does not include MSG_OOB due to its history of security
+  // vulnerabilities, see crbug.com/428177287.
+  const Arg<int> flags(argIndex);
+  return If((flags & ~(MSG_DONTWAIT | MSG_NOSIGNAL)) == 0, Allow())
+      .Else(CrashSIGSYS());
+}
+
 }  // namespace sandbox.
--- a/sandbox/linux/seccomp-bpf-helpers/syscall_parameters_restrictions.h
+++ b/sandbox/linux/seccomp-bpf-helpers/syscall_parameters_restrictions.h
@@ -126,6 +126,11 @@
 // Restrict the flags of pipe2().
 SANDBOX_EXPORT bpf_dsl::ResultExpr RestrictPipe2();
 
+// Restrict the flags of send(), sendfrom(), sendmsg(), and sendmmsg(), syscalls
+// that send a message to a socket. The flags are allowlisted, but in
+// particular, this denies MSG_OOB.
+SANDBOX_EXPORT bpf_dsl::ResultExpr RestrictSockSendFlags(int sysno);
+
 }  // namespace sandbox.
 
 #endif  // SANDBOX_LINUX_SECCOMP_BPF_HELPERS_SYSCALL_PARAMETERS_RESTRICTIONS_H_
--- a/sandbox/linux/seccomp-bpf-helpers/syscall_sets.cc
+++ b/sandbox/linux/seccomp-bpf-helpers/syscall_sets.cc
@@ -618,15 +618,6 @@
 #if defined(__i386__) || defined(__arm__) || defined(__mips__)
     case __NR__newselect:
 #endif
-#if defined(__arm__) || \
-    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
-    case __NR_send:
-#endif
-#if defined(__x86_64__) || defined(__arm__) || defined(__mips__) || \
-    defined(__aarch64__)
-    case __NR_sendmsg:  // Could specify destination.
-    case __NR_sendto:   // Could specify destination.
-#endif
     case __NR_write:
     case __NR_writev:
       return true;
@@ -639,10 +630,37 @@
     (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
     case __NR_recvmmsg_time64:  // Could specify source.
 #endif
-    case __NR_sendmmsg:  // Could specify destination.
     case __NR_splice:
     case __NR_tee:
     case __NR_vmsplice:
+// send* syscalls need their flags filtered.
+#if defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+    case __NR_send:
+#endif
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || \
+    defined(__mips__) || defined(__aarch64__)
+    case __NR_sendmsg:  // Could specify destination.
+    case __NR_sendto:   // Could specify destination.
+#endif
+    case __NR_sendmmsg:  // Could specify destination.
+    default:
+      return false;
+  }
+}
+
+bool SyscallSets::IsSockSendOneMsg(int sysno) {
+  switch (sysno) {
+#if defined(__arm__) || \
+    (defined(ARCH_CPU_MIPS_FAMILY) && defined(ARCH_CPU_32_BITS))
+    case __NR_send:
+#endif
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || \
+    defined(__mips__) || defined(__aarch64__)
+    case __NR_sendmsg:  // Could specify destination.
+    case __NR_sendto:   // Could specify destination.
+#endif
+      return true;
     default:
       return false;
   }
--- a/sandbox/linux/seccomp-bpf-helpers/syscall_sets.h
+++ b/sandbox/linux/seccomp-bpf-helpers/syscall_sets.h
@@ -58,6 +58,7 @@
 
   static bool IsAllowedAddressSpaceAccess(int sysno);
   static bool IsAllowedGeneralIo(int sysno);
+  static bool IsSockSendOneMsg(int sysno);
   static bool IsPrctl(int sysno);
   static bool IsSeccomp(int sysno);
   static bool IsAllowedBasicScheduler(int sysno);
--- a/sandbox/policy/linux/bpf_network_policy_linux.cc
+++ b/sandbox/policy/linux/bpf_network_policy_linux.cc
@@ -280,6 +280,8 @@
     case __NR_connect:
     case __NR_bind:
     case __NR_getsockname:
+      // TODO(crbug.com/40052246): restrict sendmmsg() for the network service,
+      // probably with RestrictSockSendFlags().
     case __NR_sendmmsg:
       return Allow();
     case __NR_socket:
